File: 09-HASHING/PHP/lru-cache.php
<?php

// https://leetcode.com/problems/lru-cache/
final class Node
{
    public function __construct(
        public int   $k,
        public int   $v,
        public ?Node $next = null,
        public ?Node $prev = null,
    )
    {
    }
}

class LRUCache
{
    /**
     * @var Node[]
     */
    private array $data = [];

    private ?Node $head = null;

    private ?Node $tail = null;

    function __construct(private readonly int $capacity)
    {
    }

    public function get(int $key): int
    {
        if ($node = $this->data[$key] ?? false) {
            $this->moveToTail($node);
            return $node->v;
        }

        return -1;
    }

    public function put(int $key, int $value): void
    {
        if ($node = $this->data[$key] ?? false) {
            $node->v = $value;
            $this->moveToTail($node);
            return;
        }

        $node = new Node(k: $key, v: $value, prev: $this->tail);
        $this->data[$key] = $node;
        if ($this->head === null) {
            $this->head = $node;
            $this->tail = $node;
            return;
        }

        $this->tail->next = $node;
        $this->tail = $node;
        if (count($this->data) > $this->capacity) {
            $oldHead = $this->head;
            $this->head = $this->head->next;
            $this->head->prev = null;
            unset($this->data[$oldHead->k]);
        }
    }

    private function moveToTail(Node $node): void
    {
        if ($node === $this->tail) {
            return;
        }

        if ($this->head === $node) {
            $this->head = $this->head->next;
        } else {
            $node->next->prev = $node->prev;
            $node->prev->next = $node->next;
        }

        $node->next = null;
        $node->prev = $this->tail;
        $this->tail->next = $node;
        $this->tail = $node;
    }
}

File: 09-HASHING/Python/lru-cache.py

#https://leetcode.com/problems/lru-cache/
class Node:
    def __init__(self, k: int, v: int, next: 'Node' = None, prev: 'Node' = None):
        self.k = k
        self.v = v
        self.next = next
        self.prev = prev

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.data = {}
        self.head = None
        self.tail = None

    def get(self, key: int) -> int:
        if key in self.data:
            node = self.data[key]
            self._move_to_tail(node)
            return node.v
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.data:
            node = self.data[key]
            node.v = value
            self._move_to_tail(node)
            return

        node = Node(k=key, v=value, prev=self.tail)
        self.data[key] = node

        if self.head is None:
            self.head = node
            self.tail = node
            return

        self.tail.next = node
        self.tail = node

        if len(self.data) > self.capacity:
            old_head = self.head
            self.head = self.head.next
            self.head.prev = None
            del self.data[old_head.k]

    def _move_to_tail(self, node: Node) -> None:
        if node == self.tail:
            return

        if self.head == node:
            self.head = self.head.next
        else:
            node.next.prev = node.prev
            node.prev.next = node.next

        node.next = None
        node.prev = self.tail
        self.tail.next = node
        self.tail = node